library/core/class.controller.php

```diff
@@ -1470,6 +1470,24 @@ public function renderData($Data = null) {
 
         $this->sendHeaders();
 
+        // Recursively walk the data, decoding any IP address fields.
+        $ipWalk = function(&$inputValues) use (&$ipWalk) {
+            if (is_array($inputValues) || is_object($inputValues)) {
+                foreach ($inputValues as $key => &$val) {
+                    if (is_string($val) && stringEndsWith($key, 'IPAddress', true)) {
+                        $val = ipDecode($val);
+                    } elseif (is_array($val) && stringEndsWith($key, 'IPAddresses', true)) {
+                        array_walk($val, function(&$ip, $i) {
+                            $ip = ipDecode($ip);
+                        });
+                    } elseif (is_array($val) || is_object($val)) {
+                        $ipWalk($val);
+                    }
+                }
+            }
+        };
+        $ipWalk($Data);
+
```

Comment:
My preference would be for this to be put into a function/method that could be unit tested.